File: java/diamond/src/main/java/DiamondPrinter.java
import java.util.ArrayList;
import java.util.List;
import java.util.Collections;

class DiamondPrinter {

    List<String> printToList(char a) {
        List<String> res = new ArrayList<String>();
        int diff = a - 'A';

        for (int i = 0; i <= diff; i++) {
            StringBuilder s1 = new StringBuilder();
            // build first half of line
            s1.append(spaces(diff - i) + (char)('A' + i) + spaces(i));
            StringBuilder s2 = new StringBuilder(s1);
            s1.append(s2.reverse().substring(1));
            res.add(s1.toString());
        }

        // quit now if only 'A'
        if (a == 'A') return res;

        // create bottom half by reflecting top half
        List<String> bottom = new ArrayList<String>(res);
        Collections.reverse(bottom);
        res.addAll(bottom.subList(1, bottom.size()));
        return res;
    }

    private String spaces(int n) {
        StringBuilder res = new StringBuilder();
        for (int i = 0; i < n; i++) {
            res.append(' ');
        }
        return res.toString();
    }

}

File: kotlin/diamond/src/main/kotlin/DiamondPrinter.kt
class DiamondPrinter {
  fun printToList(c: Char): List<String> {
    var res = arrayListOf<String>()
    var letters = ('A'..c).toList()
    var lineLength = letters.size * 2 - 1

    // build top half
    for ((i, letter) in letters.withIndex()) {
      var outerSpacesCount = letters.size - 1 - i;
      var outer = " ".repeat(outerSpacesCount)
      if (letter == 'A') {
        res.add("$outer$letter$outer")
      } else {
        var innerSpacesCount = lineLength - (2 + (2 * outerSpacesCount))
        var inner = " ".repeat(innerSpacesCount)
        res.add("$outer$letter$inner$letter$outer")
      }
    }
    return res + res.reversed().slice(1..res.lastIndex)
  }
}